File: python-design-patterns/functional_patterns/composite/composite.py
class Node:
    def __init__(self, value):
        self.value = value
        self.children = []

    def add(self, child):
        self.children.append(child)

    def remove(self, child):
        self.children.remove(child)

    def operation(self):
        return self.value

def sum_of_paths(tree):
    if not tree.children:
        return tree.operation()
    return tree.operation() + sum(sum_of_paths(child) for child in tree.children)


def main():
    root = Node(10)
    root.add(Node(20))
    subtree = Node(30)
    subtree.add(Node(10))
    root.add(subtree)

    print(sum_of_paths(root))


if __name__ == "__main__":
    main()

File: scala-design-patterns/src/main/scala/FunctionalPatterns/Composite/Composite.scala
import scala.collection.mutable.ListBuffer

class Node (value: Int) {
    val children = new ListBuffer[Node]()
    def +=(child: Node): Unit = children += child
    def -=(child: Node): Unit = children -= child
    def operation(): Int = this.value
}

def sum_of_paths(tree: Node): Int = {
    if (tree.children.isEmpty)
        return tree.operation()
    return tree.operation() + tree.children.map(sum_of_paths(_)).sum
}

object CompositeFunc {
    def main(args: Array[String]): Unit = {
        var root = Node(10)
        root += Node(20)
        var subtree = Node(30)
        subtree += Node(10)
        root += subtree

        println(sum_of_paths(root))
    }
}
